crates/polars-plan/src/dsl/string.rs

```diff
@@ -63,10 +63,35 @@ impl StringNameSpace {
 
     #[cfg(feature = "extract_groups")]
     // Extract all captures groups from a regex pattern as a struct
-    pub fn extract_groups(self, pat: &str) -> Expr {
-        let pat = pat.to_string();
-        self.0
-            .map_private(StringFunction::ExtractGroups { pat }.into())
+    pub fn extract_groups(self, pat: &str) -> PolarsResult<Expr> {
```

Comment:
determining the data type here, means we only do that computation of the data type once. 

The regex is compiled twice, because we cannot send the regex in the logical plan.

In the previous commit we would compile the regex every time we ask the schema, which we do a lot.

Reply:
Thanks @ritchie46 - makes sense once you see it.

I think one of the issues I was having was not understanding the order of operations fully.

`dsl/string.rs` -> `function_expr/strings.rs`

So I wasn't able to figure out at what point one could manipulate the dtype (other than what I initially tried)